### PR TITLE
quiche: Implement http2_reconstruct_object_impl.h.

### PR DESCRIPTION
**Pull request by wu-bin**
Description:

Implement http2_reconstruct_object_impl.h, which provides simple test-only apis to re-construct c++ objects in place.  Also added BUILD rule "http2_test_tools_random" since it's needed by http2_reconstruct_object_impl.h.

Risk Level: none, code not used yet.
Testing:

bazel test --test_output=all test/extensions/quic_listeners/quiche/platform:all @com_googlesource_quiche//:all

Docs Changes: none
Release Notes: none

### CODE REVIEW

File: source/extensions/quic_listeners/quiche/platform/quic_logging_impl.h [hunk 1]
```diff
@@ -59,6 +59,13 @@
 #define CHECK(condition)                                                                           \
   QUIC_LOG_IF_IMPL(FATAL, ABSL_PREDICT_FALSE(!(condition))) << "CHECK failed: " #condition "."
 
+#define CHECK_GT(a, b) CHECK((a) > (b))
+#define CHECK_GE(a, b) CHECK((a) >= (b))
+#define CHECK_LT(a, b) CHECK((a) < (b))
+#define CHECK_LE(a, b) CHECK((a) <= (b))
+#define CHECK_NE(a, b) CHECK((a) != (b))
+#define CHECK_EQ(a, b) CHECK((a) == (b))
+
```
Comment (danzh2010) [inline source/extensions/quic_listeners/quiche/platform/quic_logging_impl.h]: This is also added in #6650

Comment (wu-bin) [reply source/extensions/quic_listeners/quiche/platform/quic_logging_impl.h]: Yeah, I copied it from there:)  I copied it to the exact same location so it shouldn't cause merge conflicts.

File: test/extensions/quic_listeners/quiche/platform/http2_platform_test.cc [hunk 2]
```diff
@@ -77,16 +79,30 @@ TEST(Http2PlatformTest, Http2Log) {
   HTTP2_DLOG_EVERY_N(ERROR, 2) << "DLOG_EVERY_N(ERROR, 2)";
 }
 
+TEST(Http2PlatformTest, Http2MakeUnique) {
```
[the remark this answers is not in the document]
Comment (wu-bin) [reply test/extensions/quic_listeners/quiche/platform/http2_platform_test.cc]: Not adding this test, I just moved it up to sort the tests by name alphabetically.

[the remark this answers is not in the document]
Comment (danzh2010) [reply test/extensions/quic_listeners/quiche/platform/http2_platform_test.cc]: Oh, right. It was there. Thanks for doing this!

**Pull request merged**